library/ecdh.c

```diff
@@ -102,53 +181,103 @@ void mbedtls_ecdh_free( mbedtls_ecdh_context *ctx )
     if( ctx == NULL )
         return;
 
-    mbedtls_ecp_group_free( &ctx->grp );
-    mbedtls_ecp_point_free( &ctx->Q   );
-    mbedtls_ecp_point_free( &ctx->Qp  );
-    mbedtls_ecp_point_free( &ctx->Vi  );
-    mbedtls_ecp_point_free( &ctx->Vf  );
-    mbedtls_mpi_free( &ctx->d  );
-    mbedtls_mpi_free( &ctx->z  );
-    mbedtls_mpi_free( &ctx->_d );
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
+    mbedtls_ecdh_free_internal( ctx );
+#else
+    switch( ctx->var )
+    {
+        case MBEDTLS_ECDH_VARIANT_MBEDTLS_2_0:
+            mbedtls_ecdh_free_internal( ctx );
+            break;
+        default:
+            break;
+    }
+#endif
 }
 
-/*
- * Setup and write the ServerKeyExhange parameters (RFC 4492)
- *      struct {
- *          ECParameters    curve_params;
- *          ECPoint         public;
- *      } ServerECDHParams;
- */
-int mbedtls_ecdh_make_params( mbedtls_ecdh_context *ctx, size_t *olen,
-                      unsigned char *buf, size_t blen,
-                      int (*f_rng)(void *, unsigned char *, size_t),
-                      void *p_rng )
+static int mbedtls_ecdh_make_params_internal( mbedtls_ecdh_context *ctx,
+                                              size_t *olen,
+                                              unsigned char *buf, size_t blen,
+                                              int (*f_rng)(void *,
+                                                           unsigned char *,
+                                                           size_t),
+                                              void *p_rng )
 {
     int ret;
     size_t grp_len, pt_len;
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
+    mbedtls_ecdh_context *real_ctx = ctx;
```

Comment:
I'd prefer to change the naming: `ctx_arg` and `ctx` instead of `ctx` and `real_ctx`. This is only a suggestion, I don't insist on it, as long as it's consistent between functions.

Reply:
I wasn't happy with the `real_ctx` either, but couldn't come up with anything better.